README.md

```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
+
+```yml
+STEPS_DIR := tutorial-steps
+DEV_OUTPUTS_DIR := $(STEPS_DIR)/dev/outputs
+TEMP_DIR := $(shell mktemp -d)
+
+tutorial:
+  rm -r $(DEV_OUTPUTS_DIR) || true
+  mkdir $(DEV_OUTPUTS_DIR)
+  harness-runner ../../../../../_data/harnesses/dynamic-output-topic/kstreams.yml $(TEMP_DIR)
+  diff --strip-trailing-cr $(STEPS_DIR)/dev/expected-output.json $(DEV_OUTPUTS_DIR)/actual-output.json
+  diff --strip-trailing-cr $(STEPS_DIR)/dev/expected-special-output.json $(DEV_OUTPUTS_DIR)/actual-special-order-output.json
+```
+
+
+
+
+#### 1. Structure
+
+Three top-level sections make up the harness file:
+
+* dev - the setup and teaching portion of the tutorial (required)
+* test - test setup and execution of tests, if any (optional)
+* prod - steps to build and deploy a docker image of the tutorial code (optional)
+
+In some cases, having a test and or prod section doesn't make sense, so those sections can be left out of the harness file.  The Apache Kafka console producer and consumer tutorials are a good example of tutorials that don't need a test or prod section.
+
+The `dev`, `test`, and `prod` sections contain a top-level element `steps`.  The `steps` contains any number of well, steps for the user to walk through.  Addtionally the `harness_runner` script follows the same steps for executing the tutorial automatically during builds.  The steps in any section are structured in the same way, so we'll only discuss the make-up of a single section.
+
+For reference here is an example section
+
+```yml
+ - title: Get Confluent Platform
+      content:
+        - action: make_file
+          file: docker-compose.yml
+          render:
+            file: tutorials/console-consumer-primitive-keys-values/kafka/markup/dev/make-docker-compose.adoc
+
+        - action: execute_async
+          file: tutorial-steps/dev/docker-compose-up.sh
+          render:
+            file: tutorials/console-consumer-primitive-keys-values/kafka/markup/dev/start-compose.adoc
+
+        - action: execute
+          file: tutorial-steps/dev/wait-for-containers.sh
+          render:
+            skip: true
+
+        - name: wait for ksqldb-server and connectors
+          action: sleep
+          ms: 30000
+          render:
+            skip: true
+```
+
+
+* Title - each section starts with a `title` element and as the label suggests, the text provided here is the text used the label the step for the tutorial user and the output to the console by the harness runner.  The `title` section contains one element - `content`
+
+* content - the `content` section (a YAML `dictionary`) contains an arbitrary sized list of YAML dictionaries named `action`. An `action` key creates an anonymous step i.e. not specified in the output of the test runner.  For output in the test runner, you can provide a `-name` key followed by some text for console output.
+* action - `action` keys drive the behavior of the harness. An action key can be one of these values
```

Comment:
I noticed this section does not discuss `action: docker_ksql_cli_session`. This type probably deserves its own description below.

Reply:
ack, will do!